core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java

```diff
@@ -385,29 +405,428 @@ public synchronized void close()
+            }
+
+            long pagesRetainedSizeInBytes = getRetainedSizeInBytes(pages);
+            if (exchangeSink == null && pageBufferRetainedSizeInBytes + pagesRetainedSizeInBytes <= pageBufferCapacityInBytes) {
+                pageBuffer.putAll(taskId, pages);
+                pageBufferRetainedSizeInBytes += pagesRetainedSizeInBytes;
+                bufferedPageCount += pages.size();
+                return;
+            }
+
+            if (exchangeSink == null) {
+                verify(exchangeManager == null, "exchangeManager is not expected to be initialized");
+                verify(exchange == null, "exchange is not expected to be initialized");
+                verify(sinkInstanceHandle == null, "sinkInstanceHandle is not expected to be initialized");
+                verify(writeBuffer == null, "writeBuffer is not expected to be initialized");
+
+                try {
+                    exchangeManager = exchangeManagerRegistry.getExchangeManager();
+                }
+                catch (TrinoException e) {
+                    if (EXCHANGE_MANAGER_NOT_CONFIGURED.toErrorCode().equals(e.getErrorCode())) {
+                        throw new TrinoException(EXCHANGE_MANAGER_NOT_CONFIGURED, "Exchange manager must be configured for failure recovery capabilities to be fully functional");
+                    }
+                    throw e;
+                }
+
+                exchange = exchangeManager.createExchange(new ExchangeContext(queryId, exchangeId), 1);
+
+                ExchangeSinkHandle sinkHandle = exchange.addSink(0);
+                sinkInstanceHandle = exchange.instantiateSink(sinkHandle, 0);
+                exchange.noMoreSinks();
+                exchangeSink = exchangeManager.createSink(sinkInstanceHandle, true);
+
+                writeBuffer = new DynamicSliceOutput(DEFAULT_MAX_PAGE_SIZE_IN_BYTES);
+            }
+
+            if (!pageBuffer.isEmpty()) {
+                for (Map.Entry<TaskId, List<Slice>> entry : asMap(pageBuffer).entrySet()) {
+                    writeToSink(entry.getKey(), entry.getValue());
+                }
+                pageBuffer.clear();
+                pageBufferRetainedSizeInBytes = 0;
+            }
+
+            writeToSink(taskId, pages);
+            bufferedPageCount += pages.size();
+        }
+
+        private static long getRetainedSizeInBytes(List<Slice> pages)
+        {
+            long result = 0;
+            for (Slice page : pages) {
+                result += page.getRetainedSize();
+            }
+            return result;
+        }
+
+        private void writeToSink(TaskId taskId, List<Slice> pages)
+        {
+            verify(exchangeSink != null, "exchangeSink is expected to be initialized");
+            verify(writeBuffer != null, "writeBuffer is expected to be initialized");
+            for (Slice page : pages) {
+                // wait for the sink to unblock
+                getUnchecked(exchangeSink.isBlocked());
```

Comment:
Do we need to explicitly wait here. What is the semantics of `exchangeSink.add` if sink is blocked?

Reply:
Due to inherent racy nature I would assume that the explicit wait can be omitted (e.g.: a Sink can technically become blocked in between the `isBlocked` and `add`). However it still feels like it is better to follow the contract.